Automattic/vip-workflow-plugin
# Add Editorial Metadata support - part 1

## Description

###  Editorial Metadata

This is the start of the Editorial Metadata module. It's only focused on adding this to wp-admin so one can configure what Editorial Metadata their site will have. Another PR will be coming up, that will be focused on actually making those fields useable in the block editor. In order to avoid making a large PR, and to allow better parallel progress this has been done.

The initial starting metadata fields are the same as Edit Flow, though I've added a todo asking if we should do that or come up with our own. I have shrunk the supported field list down a lot. I figured once the implementation within the block editor was done, and/or we begin opening this up for beta access we can add more in. Again, this can be modified as well as I have added a todo mentioning this.

Beyond that, I have done the rest api differently than the custom status module. The CUD endpoints give back only the modified items, and react handles the updates to the state. Given we don't know how many fields someone might create this felt like the most efficient way to do it, and it's the rest way of doing things. I'd like to make this change to the custom status rest apis as well, and perhaps think about exposing the getAll/getBy apis as well.

As for the UI, I haven't been able to make it 100% responsive for small tablets and phones but for everyone else it works well. So just a heads up when it comes to that.

#### Screenshots

The landing page without hovering

![Screenshot 2024-08-21 at 3 33 22 PM](https://github.com/user-attachments/assets/42b2e2b0-693a-4c80-8596-defdb358e9db)

The landing page after hovering, which reveals the buttons

![Screenshot 2024-08-21 at 4 06 07 PM](https://github.com/user-attachments/assets/7ee791ca-1f32-4616-b8de-26b9b6390a4f)

The CU modal

![Screenshot 2024-08-21 at 4 06 12 PM](https://github.com/user-attachments/assets/49e720b7-7ade-4408-bf10-92672ea963f9)

The Delete modal

![Screenshot 2024-08-21 at 4 06 20 PM](https://github.com/user-attachments/assets/0b6f4322-633e-4f16-b163-7b1cdc40f041)

### Shared Components

I have made a shared components folder and moved all the PHP and JS code that we share to that. This includes the success notice and delete confirmation modal used by custom status. Ideally, the create, and edit modal might get moved there too once we switch the custom status page to use modals instead of cards for it's CUD operations.

### Tests

I ended up adding a PHP Unit testing framework as well. So now we can start writing tests for our code as we work through each feature.

### Misc

I removed a lot of the unused args and methods I could find so we could reduce the remnants of Edit Flow.

## Steps to Test

- Load up the wp-admin page
- Ensure the editorial metadata menu page shows up and it's populated with the same cards as in the screenshot
- Do some CUD operations and ensure everything works as expected
- Also play around with the other parts of the plugin to ensure nothing broke with the files I moved around and renamed
